x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java

```diff
@@ -0,0 +1,155 @@
+    private final LocalRelation localRelation;
+    private List<Attribute> lazyOutput;
+
+    public Lookup(
+        Source source,
+        LogicalPlan child,
+        Expression tableName,
+        List<NamedExpression> matchFields,
+        @Nullable LocalRelation localRelation
+    ) {
+        super(source, child);
+        this.tableName = tableName;
+        this.matchFields = matchFields;
+        this.localRelation = localRelation;
+    }
+
+    public Lookup(PlanStreamInput in) throws IOException {
+        super(Source.readFrom(in), in.readLogicalPlanNode());
+        this.tableName = in.readExpression();
+        this.matchFields = in.readCollectionAsList(i -> ((PlanStreamInput) i).readNamedExpression());
+        this.localRelation = in.readBoolean() ? new LocalRelation(in) : null;
+    }
+
+    public void writeTo(PlanStreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeLogicalPlanNode(child());
+        out.writeExpression(tableName);
+        out.writeCollection(matchFields, (o, v) -> ((PlanStreamOutput) o).writeNamedExpression(v));
+        if (localRelation == null) {
+            out.writeBoolean(false);
+        } else {
+            out.writeBoolean(true);
+            localRelation.writeTo(out);
+        }
+    }
+
+    public Expression tableName() {
+        return tableName;
+    }
+
+    public List<NamedExpression> matchFields() {
+        return matchFields;
+    }
+
+    public LocalRelation localRelation() {
+        return localRelation;
+    }
+
+    public JoinConfig joinConfig() {
+        List<Expression> conditions = new ArrayList<>(matchFields.size());
+        List<Attribute> rhsOutput = Join.makeReference(localRelation.output());
+        for (NamedExpression lhs : matchFields) {
+            for (Attribute rhs : rhsOutput) {
+                if (lhs.name().equals(rhs.name())) {
+                    conditions.add(new Equals(source(), lhs, rhs));
+                    break;
+                }
+            }
+        }
+        return new JoinConfig(JoinType.LEFT, matchFields, conditions);
+    }
+
+    @Override
+    public boolean expressionsResolved() {
+        return tableName.resolved() && Resolvables.resolved(matchFields) && localRelation != null;
+    }
+
+    @Override
+    public UnaryPlan replaceChild(LogicalPlan newChild) {
+        return new Lookup(source(), newChild, tableName, matchFields, localRelation);
+    }
+
+    @Override
+    protected NodeInfo<? extends LogicalPlan> info() {
+        return NodeInfo.create(this, Lookup::new, child(), tableName, matchFields, localRelation);
+    }
+
+    @Override
+    public List<Attribute> output() {
+        if (lazyOutput == null) {
+            List<? extends NamedExpression> rightSide = localRelation != null
+                ? Join.makeNullable(Join.makeReference(localRelation.output()))
+                : matchFields;
+            lazyOutput = mergeOutputAttributes(child().output(), rightSide);
```

Comment:
Enrich has the arguments here exactly the other way around, and I think this contributes to the buggy shadowing.